cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/worker/sidecar/GCPMetadataHttpHandlerInternal.java

```diff
@@ -0,0 +1,170 @@
+import com.google.gson.GsonBuilder;
+import com.google.inject.Singleton;
+import io.cdap.cdap.common.ForbiddenException;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.common.id.Id.Namespace;
+import io.cdap.cdap.proto.BasicThrowable;
+import io.cdap.cdap.proto.codec.BasicThrowableCodec;
+import io.cdap.common.http.HttpRequests;
+import io.cdap.common.http.HttpResponse;
+import io.cdap.http.AbstractHttpHandler;
+import io.cdap.http.HttpHandler;
+import io.cdap.http.HttpResponder;
+import io.netty.handler.codec.http.DefaultHttpHeaders;
+import io.netty.handler.codec.http.HttpRequest;
+import io.netty.handler.codec.http.HttpResponseStatus;
+import java.net.URL;
+import javax.ws.rs.DELETE;
+import javax.ws.rs.GET;
+import javax.ws.rs.PUT;
+import javax.ws.rs.Path;
+import javax.ws.rs.PathParam;
+import javax.ws.rs.QueryParam;
+import javax.ws.rs.core.HttpHeaders;
+import joptsimple.internal.Strings;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Internal {@link HttpHandler} for Artifact Localizer.
+ */
+@Singleton
+@Path("/")
+public class GCPMetadataHttpHandlerInternal extends AbstractHttpHandler {
+
+  protected static final String METADATA_FLAVOR = "Metadata-Flavor";
+  protected static final String GOOGLE = "Google";
+  private static final Logger LOG = LoggerFactory.getLogger(GCPMetadataHttpHandlerInternal.class);
+  private static final Gson GSON = new GsonBuilder().registerTypeAdapter(BasicThrowable.class,
+      new BasicThrowableCodec()).create();
+  private final CConfiguration cConf;
+  private final String metadataServiceEndpoint;
+  private Namespace namespace;
+
+  public GCPMetadataHttpHandlerInternal(CConfiguration cConf) {
+    this.cConf = cConf;
+    this.metadataServiceEndpoint = cConf.get(
+        Constants.TaskWorker.METADATA_SERVICE_END_POINT);
+  }
+
+  /**
+   * Returns the status of metadata server.
+   *
+   * @param request The {@link HttpRequest}.
+   * @param responder a {@link HttpResponder} for sending response.
+   * @throws Exception if there is any error.
+   */
+  @GET
+  @Path("/")
+  public void status(HttpRequest request, HttpResponder responder) throws Exception {
+
+    // check that metadata header is present in the request.
+    if (!request.headers().contains(METADATA_FLAVOR, GOOGLE, true)) {
+      throw new ForbiddenException(
+          String.format("Request is missing required %s header. To access the metadata server, "
+              + "you must add the %s: %s header to your request.", METADATA_FLAVOR,
+              METADATA_FLAVOR, GOOGLE));
+    }
+    responder.sendStatus(HttpResponseStatus.OK,
+        new DefaultHttpHeaders().add(METADATA_FLAVOR, GOOGLE));
+  }
+
+  @GET
+  @Path("/computeMetadata/v1/instance/service-accounts/default/token")
+  public void token(HttpRequest request, HttpResponder responder,
+      @QueryParam("scopes") String scopes) throws Exception {
+
+    LOG.info("Token requested");
```

Comment:
nit: Can we make this debug log instead of info, since this is a common endpoint? Additionally, maybe it's good to log which namespace the request is made for?

Reply:
done, thanks for pointing out.